### PR TITLE
Mutex around writer access in recorder

### LINKED ISSUE #486: Database access race condition when recording
## Description

Recording rosbags with lots of topics and high message rate causes race condidtions when writing to the database.
I believe the problem is that new topic entries are written while at the same time message data from other topics is coming in.

### Expected Behavior
`ros2 bag record -a` records all topics without errors

### Actual Behavior

The problem manifests itself in about 1 in 10 runs (with the launchfile below) in one of the following three ways:
- recording seems fine, then right after hitting ctrl-C the it logs an error and exits, the bag is almost empty (I believe something gets stuck in the topic setup)
```
...
^C[INFO] [1596518181.782926826] [rclcpp]: signal_handler(signal_value=2)
[ERROR] [1596518182.245972174] [rosbag2_transport]: Failed to record: Error when processing SQL statement. SQLite error (5): database is locked
```
- segfault a few seconds after recording starts, while still adding new topics
- error "database disk image is malformed" a few seconds after recording starts, again while adding topics (this seems to be the most common)
```
...
[INFO] [1596518141.207651288] [rosbag2_transport]: Subscribed to topic '/topic57'
[ERROR] [1596518141.560762687] [rosbag2_transport]: Failed to record: Error when processing SQL statement. SQLite error (11): database disk image is malformed
```

## To Reproduce

run many publishers. The launch file below seems to work, although not as well as my robot which is publishing at higher data rate (but fewer topics).
```
from launch import LaunchDescription
from launch_ros.actions import Node

def generate_launch_description():
    return LaunchDescription([
        Node(
            package='examples_rclcpp_minimal_publisher',
            executable='publisher_lambda',
            name='pub'+str(i),
            remappings=[
                ('/topic', '/topic'+str(i))
            ]
        ) for i in range(70)
    ])
```
then run `ros2 bag record -a`

## System (please complete the following information)
- OS: macOS 10.14, ubuntu 20.4 (aarch64)
- ROS 2 Distro: Foxy
- Version: binary patch 1

## Additional context

[Opening the database](https://github.com/ros2/rosbag2/blob/6ff9d61535c59ee539d0f70f1e1a017f467254d6/rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_wrapper.cpp#L52) in `SQLITE_OPEN_FULLMUTEX` mode instead of `SQLITE_OPEN_NOMUTEX ` fixes this. I've also tried manually adding a lock around the topic entry write and the messages write, which seems to work as well.

See also #480 for a related issue. (`database_->get_last_insert_id()` is sometimes incorrect, also caused by multiple threads using the database)

### PR DESCRIPTION
**Pull request by Stapelzeiger**
The storage plugins are not designed to be thread safe. The recorder must therefore use a mutex when accessing the writer from different contexts. This happens for topic creation during discovery as well as in message callbacks from subscribed topics.

This fixes #486 and #480.

### CODE REVIEW

**Comment by dawonn-haval** [comment]
Just wanted to leave a note that this patch seems to have made an improvement on my system. Before this patch I was only able to record sensor data bag files for a 5~10 seconds before it would hang and stop recording, but now rosbag constantly runs 1~2 minutes before hanging! 

FWIW: Foxy, Ubuntu 20.04, 5 lidar + 1 camera.

**Comment by Stapelzeiger** [comment]
@Karsten1987 I added locking in the `rosbag2_cpp` writer. I've reverted the changes in recorder, since they are redundant with the `rosbag2_cpp` ones so this should solve your second point regarding the `create_subscription` in the critical zone.

Do we need a similar thread-safe interface for the reader class? I have trouble imagining a use-case where you would read the from multiple threads with the current API (in particular the `has_next()`, `read_next()` wouldn't work nicely).

File: rosbag2_cpp/include/rosbag2_cpp/writer.hpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include <string>
 #include <unordered_map>
 #include <vector>
+#include <mutex>
```
Comment (Karsten1987) [inline rosbag2_cpp/include/rosbag2_cpp/writer.hpp:22]: alpha sort

**Comment by jarledo** [comment]
Any updates on this? Our system suffers from the bug described in #486, making it almost impossible to record our topics. Would be great to have this merged.

**Comment by Karsten1987** [comment]
Thanks for the reminder, kicking off CI on this:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12823/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7780/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10537/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12768/)

**Review by Karsten1987 — APPROVED** [review]

**Comment by Karsten1987** [comment]
@Stapelzeiger thanks for the contribution

**Comment by Karsten1987** [comment]
@adamdbrw fyi

**Comment by almaslov** [comment]
Hello! It would be great to backport this fix to Foxy. Can I do it myself according to [the Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#guidelines-for-backporting-prs) ?

**Comment by emersonknapp** [comment]
Yes, by all means. If the cherry-pick works without issue then it is an easy backport.

**Pull request merged**